chains/evm/voter/voter.go

```diff
@@ -28,42 +44,167 @@ type MessageHandler interface {
+func NewVoterWithSubscription(mh MessageHandler, client ChainClient, fabric calls.TxFabric, gasPriceClient calls.GasPricer) (*EVMVoter, error) {
+	voter := &EVMVoter{
+		mh:                   mh,
+		client:               client,
+		fabric:               fabric,
+		gasPriceClient:       gasPriceClient,
+		pendingProposalVotes: make(map[common.Hash]uint8),
+	}
+
+	ch := make(chan common.Hash)
+	_, err := client.SubscribePendingTransactions(context.TODO(), ch)
+	if err != nil {
+		return nil, err
+	}
+	go voter.trackProposalPendingVotes(ch)
+
+	return voter, nil
 }
 
+// NewVoter creates EVMVoter without pending proposal subscription.
+// It is a fallback for nodes that don't support pending transaction subscription
+// and will vote on proposals that already satisfy threshold thus wasting gas.
 func NewVoter(mh MessageHandler, client ChainClient, fabric calls.TxFabric, gasPriceClient calls.GasPricer) *EVMVoter {
 	return &EVMVoter{
-		mh:             mh,
-		client:         client,
-		fabric:         fabric,
-		gasPriceClient: gasPriceClient,
+		mh:                   mh,
+		client:               client,
+		fabric:               fabric,
+		gasPriceClient:       gasPriceClient,
+		pendingProposalVotes: make(map[common.Hash]uint8),
 	}
 }
 
+// VoteProposal checks if relayer already voted and is threshold
+// satisfied and casts a vote if it isn't.
 func (v *EVMVoter) VoteProposal(m *relayer.Message) error {
 	prop, err := v.mh.HandleMessage(m)
 	if err != nil {
 		return err
 	}
-	ps, err := calls.ProposalStatus(v.client, prop)
+
+	votedByTheRelayer, err := calls.IsProposalVotedBy(v.client, v.client.RelayerAddress(), prop)
 	if err != nil {
-		return fmt.Errorf("error getting proposal: %+v status %w", prop, err)
+		return err
 	}
-	votedByTheRelayer, err := calls.IsProposalVotedBy(v.client, v.client.RelayerAddress(), prop)
+	if votedByTheRelayer {
+		return nil
+	}
+
+	shouldVote, err := v.shouldVoteForProposal(prop, 0)
 	if err != nil {
+		log.Error().Err(err)
 		return err
 	}
-	// if this relayer had not voted for proposal and proposal is in Active or Inactive status
-	// we need to vote for it
-	if !votedByTheRelayer && (ps == relayer.ProposalStatusActive || ps == relayer.ProposalStatusInactive) {
-		hash, err := calls.VoteProposal(v.client, v.fabric, v.gasPriceClient, prop)
-		log.Debug().Str("hash", hash.String()).Uint64("nonce", prop.DepositNonce).Msgf("Voted")
+
+	if !shouldVote {
+		log.Debug().Msgf("Proposal %+v already satisfies threshold", prop)
+		return nil
+	}
+
+	hash, err := calls.VoteProposal(v.client, v.fabric, v.gasPriceClient, prop)
+	if err != nil {
+		return fmt.Errorf("voting failed. Err: %w", err)
+	}
+
+	log.Debug().Str("hash", hash.String()).Uint64("nonce", prop.DepositNonce).Msgf("Voted")
+	return nil
+}
+
+// shouldVoteForProposal checks if proposal already has threshold with pending
+// proposal votes from other relayers.
+// Only works in conjuction with NewVoterWithSubscription.
```

Comment:
Please add description how exactly it works only with NewVoterWithSubscription. Because technically this description is incorrect. This function works in both scenarios, but in regular voter senario pending votes will always be 0